pkg/ccl/changefeedccl/changefeed_test.go

```diff
@@ -966,6 +966,159 @@ func TestChangefeedSchemaChangeNoBackfill(t *testing.T) {
 	}
 }
 
+// Test checkpointing when the highwater does not move due to some issues with
+// specific spans lagging behind
+func TestChangefeedLaggingSpanCheckpointing(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	rnd, _ := randutil.NewPseudoRand()
+
+	testFn := func(t *testing.T, db *gosql.DB, f cdctest.TestFeedFactory) {
+		sqlDB := sqlutils.MakeSQLRunner(db)
+
+		knobs := f.Server().(*server.TestServer).Cfg.TestingKnobs.
+			DistSQL.(*execinfra.TestingKnobs).
+			Changefeed.(*TestingKnobs)
+
+		// Initialize table
+		sqlDB.Exec(t, `CREATE TABLE foo(val INT PRIMARY KEY)`)
+		sqlDB.Exec(t, `INSERT INTO foo (val) SELECT * FROM generate_series(1, 1000)`)
+
+		fooDesc := catalogkv.TestingGetTableDescriptor(
+			f.Server().DB(), keys.SystemSQLCodec, "d", "foo")
+		tableSpan := fooDesc.PrimaryIndexSpan(keys.SystemSQLCodec)
+
+		// In order to simulate a span that will lag behind, piggy-back off of the
+		// backfill MaxSpanRequestKeys to isolate some subspan of keys in the tablespan
+		var middleSpan roachpb.Span
+		knobs.FeedKnobs.BeforeScanRequest = func(b *kv.Batch) {
+			b.Header.MaxSpanRequestKeys = 300
+		}
+		knobs.ShouldSkipResolved = func(r *jobspb.ResolvedSpan) bool {
+			if !r.Span.Key.Equal(tableSpan.Key) && !r.Span.EndKey.Equal(tableSpan.EndKey) {
+				middleSpan = r.Span
```

Comment:
This code is probably racy -- you write middleSpan on one go routine (one emitting resolved spans), while you read it in this test.  You probably want to create e.g. 
```
middleSpanCh := make(chan roachpdb.Span, 1)
....
select {
   case middleSpanCh <- *r:
   default:
}
```
and then read that single middle span off the channel when you need it.